/**
 * The MIT License (MIT)
 * Copyright (c) 2016-2017 Intel Corporation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of 
 * this software and associated documentation files (the "Software"), to deal in 
 * the Software without restriction, including without limitation the rights to 
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of 
 * the Software, and to permit persons to whom the Software is furnished to do so, 
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all 
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS 
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR 
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER 
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN 
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

/**
 * To build, use protocol buffer version >3.2.0 and the following command from base directory:
 * $ protoc -Isrc/resources/ --java_out=src/main/java/ src/resources/genomicsdb_export_config.proto
 */

syntax = "proto2";

import "genomicsdb_coordinates.proto";
import "genomicsdb_vid_mapping.proto";
import "genomicsdb_callsets_mapping.proto";

option java_package = "org.genomicsdb.model";
option java_outer_classname = "GenomicsDBExportConfiguration";

message GenomicsDBColumnOrIntervalList {
  repeated GenomicsDBColumnOrInterval column_or_interval_list = 1;
}

message RowRange {
  required int64 low = 1;
  required int64 high = 2;
}

message RowRangeList {
  repeated RowRange range_list = 1;
}

message SampleNameList {
  repeated string sample_name_list = 1;
}

message ExportConfiguration {
  required string workspace = 1;
  oneof array {
    string array_name = 2;
    bool generate_array_name_from_partition_bounds = 18;
  }
  required string reference_genome = 3;
  repeated GenomicsDBColumnOrIntervalList query_column_ranges = 4;
  repeated RowRangeList query_row_ranges = 5;
  repeated SampleNameList query_sample_names_lists = 22; //to match the structure of query_row_ranges
  repeated string attributes = 6;
  optional string query_filter = 23;
  optional string vcf_header_filename = 7;
  optional string vcf_output_filename = 8;
  optional string vcf_output_format = 9;
  oneof vid_mapping_info {
    string vid_mapping_file = 10;
    VidMappingPB vid_mapping = 19; 
  }
  oneof callset_mapping_info {
    string callset_mapping_file = 11;
    CallsetMappingPB callset_mapping = 20;
  }
    
  optional uint32 max_diploid_alt_alleles_that_can_be_genotyped = 12;
  optional bool index_output_VCF = 13; 
  optional bool produce_GT_field = 14;
  optional bool produce_FILTER_field = 15;
  optional bool sites_only_query = 16;
  optional bool produce_GT_with_min_PL_value_for_spanning_deletions = 17;
  optional bool scan_full = 21;
}
